File: functions/invitations/create_invitation.py
from datetime import datetime, timedelta, timezone
from firebase_admin import firestore
from flask import abort
from models.constants import Collections, InvitationFields, Status, ProfileFields
from models.data_models import Invitation
from utils.logging_utils import get_logger


def create_invitation(request) -> Invitation:
    """
    Creates a new invitation from the current user.

    This function creates a new invitation document in the invitations collection.
    The invitation will have a pending status and will expire after 1 day.

    Validates that the user doesn't already have a pending invitation.

    Args:
        request: The Flask request object containing:
                - user_id: The authenticated user's ID (attached by authentication middleware)

    Returns:
        An Invitation object representing the newly created invitation

    Raises:
        404: User profile not found
    """
    logger = get_logger(__name__)
    logger.info(f"Creating invitation for user {request.user_id}")

    current_user_id = request.user_id

    # Initialize Firestore client
    db = firestore.client()

    # Get current user's profile for name and avatar
    current_user_profile_ref = db.collection(Collections.PROFILES).document(
        current_user_id
    )
    current_user_profile_doc = current_user_profile_ref.get()

    if not current_user_profile_doc.exists:
        logger.warning(f"Current user profile {current_user_id} not found")
        abort(404, description="User profile not found")

    current_user_profile = current_user_profile_doc.to_dict()

    # Create a new invitation document
    invitation_ref = db.collection(Collections.INVITATIONS).document()

    # Set expiration time (1 day from now)
    current_time = datetime.now(timezone.utc)
    expires_at = current_time + timedelta(days=1)

    # Create invitation data
    invitation_data = {
        InvitationFields.SENDER_ID: current_user_id,
        InvitationFields.USERNAME: current_user_profile.get(ProfileFields.USERNAME, ""),
        InvitationFields.NAME: current_user_profile.get(ProfileFields.NAME, ""),
        InvitationFields.AVATAR: current_user_profile.get(
            ProfileFields.AVATAR, ""
        ),
        InvitationFields.STATUS: Status.PENDING,
        InvitationFields.CREATED_AT: current_time,
        InvitationFields.EXPIRES_AT: expires_at,
    }

    # Set the invitation document
    invitation_ref.set(invitation_data)

    logger.info(f"Created invitation with ID {invitation_ref.id}")

    # Return the invitation object
    return Invitation(
        invitation_id=invitation_ref.id,
        created_at=current_time.isoformat() + "Z",
        expires_at=expires_at.isoformat() + "Z",
        sender_id=current_user_id,
        status=Status.PENDING,
        username=current_user_profile.get(ProfileFields.USERNAME, ""),
        name=current_user_profile.get(ProfileFields.NAME, ""),
        avatar=current_user_profile.get(ProfileFields.AVATAR, ""),
    )

File: functions/src/invitations/create-invitation.ts
import { Request } from 'express';
import { getFirestore, Timestamp } from 'firebase-admin/firestore';
import {
  ApiResponse,
  EventName,
  InviteEventParams,
} from '../models/analytics-events.js';
import {
  Collections,
  InvitationFields,
  ProfileFields,
  Status,
} from '../models/constants.js';
import { Invitation } from '../models/data-models.js';
import { BadRequestError } from '../utils/errors.js';
import { hasReachedCombinedLimit } from '../utils/friendship-utils.js';
import { formatInvitation } from '../utils/invitation-utils.js';
import { getLogger } from '../utils/logging-utils.js';
import { getProfileDoc, hasLimitOverride } from '../utils/profile-utils.js';

import { fileURLToPath } from 'url';
import path from 'path';

const __filename = fileURLToPath(import.meta.url);
const logger = getLogger(path.basename(__filename));

/**
 * Creates a new invitation from the current user.
 *
 * This function creates a new invitation document in the invitations collection.
 * The invitation will have a pending status and will expire after 1 day.
 *
 * Validates that:
 * 1. The user hasn't reached the combined limit of friends and active invitations (5)
 *
 * @param req - The Express request object containing:
 *              - userId: The authenticated user's ID (attached by authentication middleware)
 *              - validated_params: The validated request body containing:
 *                - receiver_name: The name of the person being invited
 *
 * @returns An ApiResponse containing the invitation and analytics
 *
 * @throws 400: User has reached the maximum number of friends and active invitations
 * @throws 404: User profile not found
 */
export const createInvitation = async (
  req: Request,
): Promise<ApiResponse<Invitation>> => {
  const currentUserId = req.userId;
  logger.info(`Creating invitation for user ${currentUserId}`);

  // Check combined limit
  const { friendCount, activeInvitationCount, hasReachedLimit } =
    await hasReachedCombinedLimit(currentUserId);
  if (hasReachedLimit) {
    const override = await hasLimitOverride(currentUserId);
    if (!override) {
      logger.warn(
        `User ${currentUserId} has reached the maximum number of friends and active invitations`,
      );
      throw new BadRequestError(
        'You have reached the maximum number of friends and active invitations',
      );
    }
  }

  // Initialize Firestore client
  const db = getFirestore();

  // Get current user's profile for name and avatar
  const { data: currentUserProfile } = await getProfileDoc(currentUserId);

  const validatedParams = req.validated_params;

  // Create a new invitation document
  const invitationRef = db.collection(Collections.INVITATIONS).doc();

  // Set expiration time (1 day from now)
  const currentTime = Timestamp.now();
  const expiresAt = new Timestamp(
    currentTime.seconds + 24 * 60 * 60, // Add 24 hours in seconds
    currentTime.nanoseconds,
  );

  // Create invitation data
  const invitationData = {
    [InvitationFields.SENDER_ID]: currentUserId,
    [InvitationFields.USERNAME]:
      currentUserProfile[ProfileFields.USERNAME] || '',
    [InvitationFields.NAME]: currentUserProfile[ProfileFields.NAME] || '',
    [InvitationFields.AVATAR]: currentUserProfile[ProfileFields.AVATAR] || '',
    [InvitationFields.STATUS]: Status.PENDING,
    [InvitationFields.CREATED_AT]: currentTime,
    [InvitationFields.EXPIRES_AT]: expiresAt,
    [InvitationFields.RECEIVER_NAME]: validatedParams.receiver_name,
  };

  // Set the invitation document
  await invitationRef.set(invitationData);

  logger.info(`Created invitation with ID ${invitationRef.id}`);

  // Return the invitation object
  const invitation = formatInvitation(invitationRef.id, invitationData);

  // Create analytics event
  const event: InviteEventParams = {
    friend_count: friendCount,
    invitation_count: activeInvitationCount + 1, // Add 1 for the new invitation
  };

  return {
    data: invitation,
    status: 201,
    analytics: {
      event: EventName.INVITE_CREATED,
      userId: currentUserId,
      params: event,
    },
  };
};
